### ISSUE TITLE
clang++ gives false warnings about self-assign-overloaded expressions

### ISSUE THREAD

**Issue by 5c4ff85d-6f52-4e70-9cb3-0d9f272fa5c7** [llvm/llvm-project#42469]
|  |  |
| --- | --- |
| Bugzilla Link | [43124](https://llvm.org/bz43124) |
| Version | 8.0 |
| OS | Linux |
| CC | @gruech,@Quuxplusone,@dwblaikie,@LebedevRI,@riccibruno,@zygoloid,@rjmccall |

## Extended Description 
Clang 8.0.1 produces false warnings for expressions like `a -= a` and `a /= a`, even when `a` is not a number and the overloaded operators don't do any arithmetic operations, let alone when their return type is completely different from `decltype(a)`.

Consider this code:
```
class Vector {
public:
    Vector& operator+=(const Vector &v) { return *this; }
    Vector& operator-=(const Vector &v) { return *this; }
    Vector& operator*=(const Vector &v) { return *this; }
    Vector& operator/=(const Vector &v) { return *this; }
};

int main()
{
    Vector v;
    v += v;
    v -= v;
    v *= v;
    v /= v;
}
```

When compiled with clang++ 8.0.1, I get the following warnings:
```
$ clang++ -Wall example2.cpp -o example2
example2.cpp:13:7: warning: explicitly assigning value of variable of type 'Vector' to
      itself [-Wself-assign-overloaded]
    v -= v;
    ~ ^  ~
example2.cpp:15:7: warning: explicitly assigning value of variable of type 'Vector' to
      itself [-Wself-assign-overloaded]
    v /= v;
    ~ ^  ~
2 warnings generated.
```

See also https://stackoverflow.com/q/57645872 and https://github.com/pybind/pybind11/issues/1893 for a real-world non-trivial and non-arithmetic example of self-assignment expressions using overloaded operators.

**Comment by python3kgae** [comment]
Created https://reviews.llvm.org/D146897 for this.
